src/mono/mono/mini/mini-amd64.c

```diff
@@ -1473,6 +1473,12 @@ mono_arch_set_native_call_context_ret (CallContext *ccontext, gpointer frame, Mo
 		interp_cb->frame_arg_to_data ((MonoInterpFrameHandle)frame, sig, -1, storage);
 		if (temp_size)
 			arg_set_val (ccontext, ainfo, storage);
+
+		// amd64 backend expects I4 return value to be sign extended. Smaller integer types
+		// are extended in mono_emit_widen_call_res. On arm64, caller always does the sign
+		// extend for I4 in emit_move_return_value.
+		if (sig->ret->type == MONO_TYPE_I4 && ainfo->storage == ArgInIReg && ainfo->reg == AMD64_RAX)
```

Comment:
Not sure if its related, but since this is a regression, we had this change in .net9, https://github.com/dotnet/runtime/pull/105150/files#diff-825cf042446952a1a37306860e623ab492e209415f3d7405dc7f3be4add2cd0eR44, that dropped sign extending an I4 when converting from long, maybe that is what gets into play here as well?

Reply:
Sorry for the missing context. This regression was caused by https://github.com/dotnet/runtime/commit/4185f94ec85e619b7dfa82357ac853846dfda3af#diff-9b308c3808c534cb85f5321c4d923e794447ef5da713ed0c42ca37445bd4f51bL1327, which used to zero the high bits of RAX (which is not really correct since it should have been a sign extend). This fix is localized to this interpreter location. It is not clear to me if the linked change could create problems.